File: unimax_chess.py
import random
import copy

ROWS = 8
COLS = 8
P1, P2 = True, False
PAWN_VAL = 1
KNIGHT_VAL = 3
BISHOP_VAL = 3
ROOK_VAL = 5
QUEEN_VAL = 9
KING_VAL = 200
MOVE_VAL_DENOMINATOR = 10


class Piece(object):
    def __init__(self, is_player1):
        self.is_player1 = is_player1
        self.val = 0
        
    def is_empty(self, piece):
        return not piece
    
    def is_opponent(self, piece):
        return piece and piece.is_player1 != self.is_player1
    
    def is_in_bounds(self, r, c):
        return 0 <= r < ROWS and 0 <= c < COLS
    
    def is_self(self, piece):
        return piece == self
    
    def is_valid_move(self, r, c, board):
        return self.is_in_bounds(r, c) and \
            not board[r][c] == self and \
                (self.is_empty(board[r][c]) or self.is_opponent(board[r][c]))
    

class Pawn(Piece):
    def __init__(self, is_player1):
        super().__init__(is_player1)
        self.val = PAWN_VAL
    
    def get_valid_moves(self, row, col, board):
        """Player 1's pawns begins at index 1 and move positively,
        Player 2's pawns begin at index 6 and move negatively."""
        
        init_row = 1 if self.is_player1 else 6
        direction = 1 if self.is_player1 else -1

        if row == init_row and self.is_empty(board[row + 2 * direction][col]) and self.is_empty(board[row + direction][col]):
            yield (row + 2 * direction, col)
        if 0 <= row + direction < 8 and self.is_empty(board[row + direction][col]):
            yield (row + direction, col)
        if 0 < row < 7:
            if col > 0 and self.is_opponent(board[row + direction][col - 1]):
                yield (row + direction, col - 1)
            if col < 7 and self.is_opponent(board[row + direction][col + 1]):
                yield (row + direction, col + 1)

    def __str__(self):
        return "p>" if self.is_player1 else "<p"

class Rook(Piece):
    def __init__(self, is_player1):
        super().__init__(is_player1)
        self.val = ROOK_VAL
    
    def get_valid_moves(self, row, col, board):
        up = [(r, col) for r in range(row + 1, ROWS)]
        down = [(r, col) for r in range(row - 1, -1, -1)]
        left = [(row, c) for c in range(col - 1, -1, -1)]
        right = [(row, c) for c in range(col + 1, COLS)]
        paths = [up, down, left, right]

        for path in paths:
            for r, c in path:
                if self.is_empty(board[r][c]):
                    yield (r, c)
                elif self.is_opponent(board[r][c]):
                    yield (r, c)
                    break
                elif board[r][c]:
                    break

    def __str__(self):
        return "R>" if self.is_player1 else "<R"
    
class Knight(Piece):
    def __init__(self, is_player1):
        super().__init__(is_player1)
        self.val = KNIGHT_VAL
    
    def get_valid_moves(self, row, col, board):
        for r, c in [(row - 2, col - 1), (row - 2, col + 1),
                     (row + 2, col - 1), (row + 2, col + 1),
                     (row - 1, col - 2), (row - 1, col + 2),
                     (row + 1, col - 2), (row + 1, col + 2)]:
            
            if self.is_valid_move(r, c, board):
                yield (r, c)
    
    def __str__(self):
        return "N>" if self.is_player1 else "<N"

class Bishop(Piece):
    def __init__(self, is_player1):
        super().__init__(is_player1)
        self.val = BISHOP_VAL
    
    def get_valid_moves(self, row, col, board):
        diag1 = [(row + i, col + i) for i in range(1, min(ROWS - row, COLS - col))]
        diag2 = [(row + i, col - i) for i in range(1, min(ROWS - row, col + 1))]
        diag3 = [(row - i, col + i) for i in range(1, min(row + 1, COLS - col))]
        diag4 = [(row - i, col - i) for i in range(1, min(row + 1, col + 1))]

        diagonals = [diag1, diag2, diag3, diag4]

        for diagonal in diagonals:
            for r, c in diagonal:
                if self.is_empty(board[r][c]):
                    yield (r, c)
                elif self.is_opponent(board[r][c]):
                    yield (r, c)
                    break
                elif board[r][c]:
                    break

    def __str__(self):
        return "B>" if self.is_player1 else "<B"

class King(Piece):
    def __init__(self, is_player1):
        super().__init__(is_player1)
        self.val = KING_VAL
    
    def get_valid_moves(self, row, col, board):
        for r in range(row - 1, row + 2):
            for c in range(col - 1, col + 2):
                if self.is_valid_move(r, c, board):
                    yield (r, c)

    def __str__(self):
        return "K>" if self.is_player1 else "<K"

class Queen(Piece):
    def __init__(self, is_player1):
        super().__init__(is_player1)
        self.val = QUEEN_VAL
    
    def get_valid_moves(self, row, col, board):
        diag1 = [(row + i, col + i) for i in range(1, min(ROWS - row, COLS - col))]
        diag2 = [(row + i, col - i) for i in range(1, min(ROWS - row, col + 1))]
        diag3 = [(row - i, col + i) for i in range(1, min(row + 1, COLS - col))]
        diag4 = [(row - i, col - i) for i in range(1, min(row + 1, col + 1))]
        up = [(r, col) for r in range(row + 1, ROWS)]
        down = [(r, col) for r in range(row - 1, -1, -1)]
        left = [(row, c) for c in range(col - 1, -1, -1)]
        right = [(row, c) for c in range(col + 1, COLS)]
        paths = [diag1, diag2, diag3, diag4, up, down, left, right]

        for path in paths:
            for r, c in path:
                if self.is_empty(board[r][c]):
                    yield (r, c)
                elif self.is_opponent(board[r][c]):
                    yield (r, c)
                    break
                elif board[r][c]:
                    break

    def __str__(self):
        return "Q>" if self.is_player1 else "<Q"



INIT_BOARD = [[Rook(P1), Knight(P1), Bishop(P1), Queen(P1), King(P1), Bishop(P1), Knight(P1), Rook(P1)],
             [Pawn(P1) for _ in range(8)],
             [False] * 8,
             [False] * 8,
             [False] * 8,
             [False] * 8,
             [Pawn(P2) for _ in range(8)],
             [Rook(P2), Knight(P2), Bishop(P2), Queen(P2), King(P2), Bishop(P2), Knight(P2), Rook(P2)]]


class ChessGame(object):

    def __init__(self, board):
        self.is_being_played = False
        self.board = board
        self.prev_states = set()
        if self.game_over():
            self.val = 0
        else:
            self.val = sum(piece.val for row in self.board for piece in row if piece) + \
                (sum(1 for _ in self.legal_moves(True)) + sum(1 for _ in self.legal_moves(False))) / \
                    MOVE_VAL_DENOMINATOR

    def get_board(self):
        return self.board

    def reset(self):
        self.board = copy.deepcopy(INIT_BOARD)
    
    def legal_moves(self, is_player1):
        for r in range(ROWS):
            for c in range(COLS):
                piece = self.board[r][c]
                if piece and piece.is_player1 == is_player1:
                    for move_to in piece.get_valid_moves(r, c, self.board):
                        move_from = (r, c)
                        yield move_from, move_to


    def perform_move(self, move_from, move_to):
        from_row, from_col = move_from
        to_row, to_col = move_to
        captured_piece = self.board[to_row][to_col]
        if captured_piece:
            self.val -= captured_piece.val
        self.board[to_row][to_col] = self.board[from_row][from_col]
        self.board[from_row][from_col] = False

    def game_over(self):
        if sum(1 for row in self.board for piece in row if piece and piece.val == KING_VAL) < 2:
            return True
        return False

    def copy(self):
        return ChessGame([row[:] for row in self.board])
    
    def successors(self, is_player1):
        moves = list(self.legal_moves(is_player1))
        random.shuffle(moves)
        for move in moves:
            new_game = self.copy()
            new_game.perform_move(*move)
            yield (move, new_game)

    def get_random_move(self, is_player1):
        return random.choice(list(self.legal_moves(is_player1)))

    def get_best_move(self, is_player1, limit):
        root = Node(self, None, is_player1, None)
        root.search_for_value(limit, None, root)
        return root.best_move


class Node(object):
    
    def __init__(self, game_obj, parent_node, is_player1, move):
        self.game = game_obj
        self.parent = parent_node
        self.is_player1 = is_player1
        self.move = move
        self.best_move = None
        self.num_leaf_nodes = 0
        self.val = 0
        self.depth = 0 if parent_node is None else parent_node.depth + 1
      
    def is_leaf(self, limit):
        return self.depth == limit or self.game.game_over()

    def update_parent(self):
        if self.parent is not None and self.val > self.parent.val:
            self.parent.val = self.val
            self.parent.best_move = self.move

    def should_prune(self):
        if self.parent is not None and self.game.val >= self.parent.game.val:
            return True
        return False

    def search_for_value(self, limit, move, root):
        if self.is_leaf(limit):
            self.val = self.game.val
        else:
            for move, child_game in self.game.successors(self.is_player1):
                if tuple(tuple(x) for x in child_game.get_board()) in root.game.prev_states:                    
                    continue
                child_node = Node(child_game, self, not self.is_player1, move)
                child_node.search_for_value(limit, move, root)
                if self.should_prune():
                    self.update_parent()
                    return
        self.update_parent()

def create_chess_game():
    return ChessGame(INIT_BOARD)

File: js/unimax_chess.js
// Constants
const ROWS = 8;
const COLS = 8;
const P1 = true;
const P2 = false;
const PAWN_VAL = 1;  // initial values from https://www.pi.infn.it/%7Ecarosi/chess/shannon.txt
const KNIGHT_VAL = 3;
const BISHOP_VAL = 3;
const ROOK_VAL = 5;
const QUEEN_VAL = 9;
const KING_VAL = 200;
const MOVE_VAL_DENOMINATOR = 9; // ~ between preferring no capture vs preferring free movement


class Piece {
    constructor(isPlayer1) {
        this.isPlayer1 = isPlayer1;
        this.val = 0;
    }

    isEmpty(piece) {
        return !piece;
    }

    isOpponent(piece) {
        return piece && piece.isPlayer1 !== this.isPlayer1;
    }

    isInBounds(r, c) {
        return 0 <= r && r < ROWS && 0 <= c && c < COLS;
    }

    isSelf(piece) {
        return piece === this;
    }

    isValidMove(r, c, board) {
        return (
            this.isInBounds(r, c) &&
            board[r][c] !== this &&
            (this.isEmpty(board[r][c]) || this.isOpponent(board[r][c]))
        );
    }

    getValidMoves(row, col, board) {
        return [];
    }
}


class Pawn extends Piece {
    constructor(isPlayer1) {
        super(isPlayer1);
        this.val = PAWN_VAL;
    }

    getValidMoves(row, col, board) {
        /**
         * Player 1's pawn begins at index 1 and moves positively,
         * Player 2's pawn begins at index 6 and moves negatively.
         */
        const moves = [];
        const initRow = this.isPlayer1 ? 1 : 6;
        const direction = this.isPlayer1 ? 1 : -1;

        // check initial double forward move
        if (
            row == initRow &&
            this.isEmpty(board[row + 2 * direction][col]) &&
            this.isEmpty(board[row + direction][col])
        ) {
            moves.push([row + 2 * direction, col]);
        }

        // check single forward move
        if (
            0 <= row + direction && row + direction < ROWS &&
            this.isEmpty(board[row + direction][col])
        ) {
            moves.push([row + direction, col]);
        }

        // check diagonal captures
        if (0 < row && row < ROWS - 1) {
            if (col > 0 && this.isOpponent(board[row + direction][col - 1])) {
              moves.push([row + direction, col - 1]);
            }
            if (col < COLS - 1 && this.isOpponent(board[row + direction][col + 1])) {
              moves.push([row + direction, col + 1]);
            }
        }
        return moves;
    }

    toString() {
        return this.isPlayer1 ? '♙' : '♟';
    }
}


class Rook extends Piece {
    constructor(isPlayer1) {
        super(isPlayer1);
        this.val = ROOK_VAL;
    }

    getValidMoves(row, col, board) {
        const moves = [];
        const up = Array.from({ length: ROWS - row - 1 }, (_, i) => [row + i + 1, col]);
        const down = Array.from({ length: row }, (_, i) => [row - i - 1, col]);
        const left = Array.from({ length: col }, (_, i) => [row, col - i - 1]);
        const right = Array.from({ length: COLS - col - 1 }, (_, i) => [row, col + i + 1]);
        const paths = [up, down, left, right];

        for (const path of paths) {
            for (const [r, c] of path) {
                if (this.isEmpty(board[r][c])) {
                    moves.push([r, c]);
                } else if (this.isOpponent(board[r][c])) {
                    moves.push([r, c]);
                    break; // Stop after capturing an opponent piece
                } else {
                    break; // Stop at any other obstruction
                }
            }
        }

        return moves;
    }

    toString() {
        return this.isPlayer1 ? '♖' : '♜';
    }
}


class Knight extends Piece {
    constructor(isPlayer1) {
        super(isPlayer1);
        this.val = KNIGHT_VAL;
    }

    getValidMoves(row, col, board) {
        const moves = [];
        const knightMoves = [
            [row - 2, col - 1], [row - 2, col + 1],
            [row + 2, col - 1], [row + 2, col + 1],
            [row - 1, col - 2], [row - 1, col + 2],
            [row + 1, col - 2], [row + 1, col + 2],
        ];

        for (const [r, c] of knightMoves) {
            if (this.isValidMove(r, c, board)) {
              moves.push([r, c]);
            }
        }
      
        return moves;
    }

    toString() {
        return this.isPlayer1 ? '♘' : '♞';
    }
}


class Bishop extends Piece {
    constructor(isPlayer1) {
        super(isPlayer1);
        this.val = BISHOP_VAL;
    }

    getValidMoves(row, col, board) {
        const moves = [];
        const diag1 = Array.from(
            { length: Math.min(ROWS - row, COLS - col) - 1 },
            (_, i) => [row + i + 1, col + i + 1]
        );
        const diag2 = Array.from(
            { length: Math.min(ROWS - row, col + 1) - 1 },
            (_, i) => [row + i + 1, col - i - 1]
        );
        const diag3 = Array.from(
            { length: Math.min(row + 1, COLS - col) - 1 },
            (_, i) => [row - i - 1, col + i + 1]
        );
        const diag4 = Array.from(
            { length: Math.min(row + 1, col + 1) - 1 },
            (_, i) => [row - i - 1, col - i - 1]
        );
      
        const diagonals = [diag1, diag2, diag3, diag4];

        for (const diagonal of diagonals) {
            for (const [r, c] of diagonal) {
                if (this.isEmpty(board[r][c])) {
                    moves.push([r, c]);
                } else if (this.isOpponent(board[r][c])) {
                    moves.push([r, c]);
                    break; // Stop after capturing an opponent piece
                } else {
                    break; // Stop at any other obstruction
                }
            }
        }
      
        return moves;
    }

    toString() {
        return this.isPlayer1 ? '♗' : '♝';
    }
}


class King extends Piece {
    constructor(isPlayer1) {
        super(isPlayer1);
        this.val = KING_VAL;
    }

    getValidMoves(row, col, board) {
        const moves = [];
        for (let r = row - 1; r <= row + 1; r++) {
            for (let c = col - 1; c <= col + 1; c++) {
                if (this.isValidMove(r, c, board)) {
                    moves.push([r, c]);
                }
            }
        }
        return moves;
    }

    toString() {
        return this.isPlayer1 ? '♔' : '♚';
    }
}


class Queen extends Piece {
    constructor(isPlayer1) {
        super(isPlayer1);
        this.val = QUEEN_VAL;
    }

    getValidMoves(row, col, board) {
        const moves = [];

        // Define diagonal paths
        const diag1 = Array.from(
            { length: Math.min(ROWS - row, COLS - col) - 1 },
            (_, i) => [row + i + 1, col + i + 1]
        );
        const diag2 = Array.from(
            { length: Math.min(ROWS - row, col + 1) - 1 },
            (_, i) => [row + i + 1, col - i - 1]
        );
        const diag3 = Array.from(
            { length: Math.min(row + 1, COLS - col) - 1 },
            (_, i) => [row - i - 1, col + i + 1]
        );
        const diag4 = Array.from(
            { length: Math.min(row + 1, col + 1) - 1 },
            (_, i) => [row - i - 1, col - i - 1]
        );

        // Define straight paths
        const up = Array.from(
            { length: ROWS - row - 1 },
            (_, i) => [row + i + 1, col]
        );
        const down = Array.from(
            { length: row },
            (_, i) => [row - i - 1, col]
        );
        const left = Array.from(
            { length: col },
            (_, i) => [row, col - i - 1]
        );
        const right = Array.from(
            { length: COLS - col - 1 },
            (_, i) => [row, col + i + 1]
        );

        const paths = [diag1, diag2, diag3, diag4, up, down, left, right];

        for (const path of paths) {
            for (const [r, c] of path) {
                if (this.isEmpty(board[r][c])) {
                    moves.push([r, c]);
                } else if (this.isOpponent(board[r][c])) {
                    moves.push([r, c]);
                    break; // Stop after capturing an opponent piece
                } else {
                    break; // Stop at any other obstruction
                }
            }
        }

        return moves;
    }

    toString() {
        return this.isPlayer1 ? '♕' : '♛';
    }
}


const INIT_BOARD = [
    [new Rook(P1), new Knight(P1), new Bishop(P1), new Queen(P1), new King(P1), new Bishop(P1), new Knight(P1), new Rook(P1)],
    Array(8).fill(null).map(() => new Pawn(P1)),
    Array(8).fill(null),
    Array(8).fill(null),
    Array(8).fill(null),
    Array(8).fill(null),
    Array(8).fill(null).map(() => new Pawn(P2)),
    [new Rook(P2), new Knight(P2), new Bishop(P2), new Queen(P2), new King(P2), new Bishop(P2), new Knight(P2), new Rook(P2)]
];


class ChessGame {
    constructor(board = INIT_BOARD) {
        this.isBeingPlayed = false;
        this.board = board;
        this.prevStates = new Set();  // so as to not move to any previously visited state
        this.val = this.isGameOver() ? 0 : this.getUnimaxScore();
    }

    getBoard() {
        return this.board;
    }

    getUnimaxScore() {
        /**
         * unimax = the sum of the values of all pieces on the board plus
         * the number of legal moves available to both players divided by MOVE_VAL_DENOMINATOR
         * A higher piece value perfers states that keep that piece on the board
         * A lower move denominator prefers states that have more legal moves available
         * (sometimes at the expense of sacreficing a piece)
         */
        return this.board.flat().reduce((sum, piece) => sum + (piece ? piece.val : 0), 0) +
        (this.countLegalMoves(true) + this.countLegalMoves(false)) / MOVE_VAL_DENOMINATOR;
    }

    countLegalMoves(isPlayer1) {
        let moves = 0;
        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
            const piece = this.board[r][c];
            if (piece && piece.isPlayer1 === isPlayer1) {
                moves += piece.getValidMoves(r, c, this.board).length;
            }
            }
        }
        return moves;
    }

    getLegalMoves(isPlayer1) {  // TODO: optimize?
        const moves = [];
        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
            const piece = this.board[r][c];
            if (piece && piece.isPlayer1 === isPlayer1) {
                const validMoves = piece.getValidMoves(r, c, this.board);
                for (const moveTo of validMoves) {
                moves.push({ moveFrom: [r, c], moveTo });
                }
            }
            }
        }
        return moves;
    }

    performMove(moveFrom, moveTo) {
        const [fromRow, fromCol] = moveFrom;
        const [toRow, toCol] = moveTo;
        const capturedPiece = this.board[toRow][toCol];
        if (capturedPiece) {
            this.val -= capturedPiece.val;
        }
        this.board[toRow][toCol] = this.board[fromRow][fromCol];
        this.board[fromRow][fromCol] = null;
    }

    isGameOver() {
        const kingCount = this.board.flat().filter(piece => piece && piece.val === KING_VAL).length;
        return kingCount < 2;
    }

    copy() {
        const boardCopy = [];
        for (let r = 0; r < ROWS; r++) {
            const rowCopy = [];
            for (let c = 0; c < COLS; c++) {
                const cell = this.board[r][c];
                if (cell) {
                    const pieceCopy = new cell.constructor(cell.isPlayer1);
                    rowCopy.push(pieceCopy);
                } else {
                    rowCopy.push(null);
                }
            }
            boardCopy.push(rowCopy);
        }
        return new ChessGame(boardCopy);
    }

    *successors(isPlayer1) {
        const moves = [...this.getLegalMoves(isPlayer1)];
        for (let i = moves.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [moves[i], moves[j]] = [moves[j], moves[i]];
        }
      
        for (const move of moves) {
            const newGame = this.copy();
            newGame.performMove(move.moveFrom, move.moveTo);
            yield { move, game: newGame };
        }
    }

    getRandomMove(isPlayer1) {
        const legalMoves = this.getLegalMoves(isPlayer1);
        if (legalMoves.length === 0) return null;
        const randomIndex = Math.floor(Math.random() * legalMoves.length);
        return legalMoves[randomIndex];
    }

    getBestMove(isPlayer1, limit) {
        const root = new Node(this, null, isPlayer1, null);
        root.searchForValue(limit, null, root);
        return root.bestMove;
    }
        

}


class Node {
    constructor(gameObj, parentNode, isPlayer1, move) {
        this.game = gameObj;
        this.parent = parentNode;
        this.isPlayer1 = isPlayer1;
        this.move = move;
        this.bestMove = null;
        this.val = 0;
        this.depth = parentNode === null ? 0 : parentNode.depth + 1;
    }

    isLeaf(limit) {
        return this.depth === limit || this.game.isGameOver();
    }

    updateParent() {
        if (this.parent !== null && this.val > this.parent.val) {
            this.parent.val = this.val;
            this.parent.bestMove = this.move;
        }
    }

    shouldPrune() {
        return this.parent !== null && this.game.val >= this.parent.game.val;
    }

    searchForValue(limit, move, root) {
        if (this.isLeaf(limit)) {
            this.val = this.game.val;
        } else {
            for (const { move, game: childGame } of this.game.successors(this.isPlayer1)) {
                const boardState = JSON.stringify(childGame.getBoard());
                if (root.game.prevStates.has(boardState)) {
                    continue;
                }
                const childNode = new Node(childGame, this, !this.isPlayer1, move);
                childNode.searchForValue(limit, move, root);
                if (this.shouldPrune()) {
                    this.updateParent();
                    return;
                }
            }
        }
        this.updateParent();
    }
}


function createChessGame() {
    return new ChessGame(INIT_BOARD);
}


// Initialize the game when the DOM content is loaded
document.addEventListener('DOMContentLoaded', () => {
    const game = createChessGame();
    playGame(game);
});

function initializeBoard(board) {
    const chessboardElement = document.getElementById('chessboard');

    // Create squares
    for (let row = 0; row < ROWS; row++) {
        for (let col = 0; col < COLS; col++) {
            const square = document.createElement('div');
            square.classList.add('square');

            if ((row + col) % 2 === 0) {
                square.classList.add('light-square');
            } else {
                square.classList.add('dark-square');
            }

            // Position the square
            square.style.left = `${col * 60}px`;
            square.style.top = `${row * 60}px`;

            chessboardElement.appendChild(square);
        }
    }

    // Create pieces
    const pieceElements = new Map(); // Map to keep track of piece elements

    for (let row = 0; row < ROWS; row++) {
        for (let col = 0; col < COLS; col++) {
            const piece = board[row][col];
            if (piece) {
                const pieceElement = document.createElement('div');
                pieceElement.classList.add('piece');
                pieceElement.textContent = piece.toString();

                // Position the piece
                pieceElement.style.left = `${col * 60}px`;
                pieceElement.style.top = `${row * 60}px`;

                chessboardElement.appendChild(pieceElement);

                // Store the piece element for later reference
                pieceElements.set(`${row},${col}`, pieceElement);
            }
        }
    }

    return pieceElements;
}

function playGame(game) {
    const PIECE_ANIMATION_DURATION_MS = 50; // TODO: Match CSS
    const TRAIL_FADEOUT_DURATION_MS = 15000;
    const TIMEOUT_BUFFER_MS = 20;

    let isPlayer1Turn = true;
    const capturedList = [];
    let moveCount = 0;
    const chessboardElement = document.getElementById('chessboard');
    const pieceElements = initializeBoard(game.getBoard());
    const turnInfoElement = document.getElementById('turn-info');
    const capturedPiecesElement = document.getElementById('captured-pieces');


    function updateGameInfo() {
        turnInfoElement.textContent = `Turn ${moveCount + 1}`;
        capturedPiecesElement.textContent = `Captured pieces: ${capturedList.join(' ')}`;
    }

    updateGameInfo();

    function makeMove() {
        if (game.isGameOver()) {
            turnInfoElement.textContent = 'Game over';
            return;
        }

        // TODO: adjust unimax search depth. 2 = shallow/quick, 4 = deep/slow
        const bestMove = game.getBestMove(isPlayer1Turn, 3);
        if (!bestMove) {
            return;
        }

        const { moveFrom, moveTo } = bestMove;
        const capturedPiece = game.getBoard()[moveTo[0]][moveTo[1]];

        // Perform the move in the game logic
        game.performMove(moveFrom, moveTo);

        // Update the UI
        const fromKey = `${moveFrom[0]},${moveFrom[1]}`;
        const toKey = `${moveTo[0]},${moveTo[1]}`;
        const pieceElement = pieceElements.get(fromKey);

        // Handle captured piece
        if (capturedPiece) {
            const capturedPieceElement = pieceElements.get(toKey);
            if (capturedPieceElement) {
                capturedPieceElement.style.opacity = '0';
                setTimeout(() => {
                    try {
                        capturedPieceElement.remove();
                    } catch (error) {
                        console.error('Error removing captured piece:', error);
                    }
                }, PIECE_ANIMATION_DURATION_MS);
                pieceElements.delete(toKey);
            }
            capturedList.push(capturedPiece.toString());
        }

        // Update the piece's position in the map
        pieceElements.delete(fromKey);
        pieceElements.set(toKey, pieceElement);

        // Animate the piece movement by changing 'left' and 'top'
        pieceElement.style.left = `${moveTo[1] * 60}px`;
        pieceElement.style.top = `${moveTo[0] * 60}px`;


        function createAndAnimateTrail(moveFrom, moveTo) {
            const trailHeight = 60;
            const trailElement = document.createElement('div');
            trailElement.classList.add('trail');
            if (isPlayer1Turn) {
                trailElement.classList.add('trail-player1');
            } else {
                trailElement.classList.add('trail-player2');
            }
        
            const startX = moveFrom[1] * 60 + 30; // Center X of the start square
            const startY = moveFrom[0] * 60 + 30; // Center Y of the start square
            const endX = moveTo[1] * 60 + 30;     // Center X of the end square
            const endY = moveTo[0] * 60 + 30;     // Center Y of the end square
        
            // Calculate the distance and angle
            const deltaX = endX - startX;
            const deltaY = endY - startY;
            const distance = Math.hypot(deltaX, deltaY);
            const angle = Math.atan2(deltaY, deltaX);
        
            // Position the trail at the starting point
            trailElement.style.left = `${startX}px`;
            trailElement.style.top = `${startY - trailHeight / 2}px`;

            // Set the width to the distance between the two points
            trailElement.style.width = `${distance}px`;
        
            // Rotate the trail to align with the movement direction
            trailElement.style.transformOrigin = '0% 50%';
            trailElement.style.transform = `rotate(${angle}rad)`;
        
            const firstPieceElement = chessboardElement.querySelector('.piece');
            if (firstPieceElement) {
                chessboardElement.insertBefore(trailElement, firstPieceElement);
            } else {
                chessboardElement.appendChild(trailElement);
            }
        
            // Start the fade-out animation
            setTimeout(() => {
                trailElement.style.opacity = '0';
            }, TIMEOUT_BUFFER_MS); // Small delay to ensure the trail is rendered before fading
        
            // Remove the trail after the animation completes
            setTimeout(() => {
                trailElement.remove();
            }, TRAIL_FADEOUT_DURATION_MS + TIMEOUT_BUFFER_MS);
        }
        
        // Animate the piece's "trail"
        createAndAnimateTrail(moveFrom, moveTo, chessboardElement);

        moveCount++;
        isPlayer1Turn = !isPlayer1Turn;

        updateGameInfo();

        // Schedule the next move after the animation completes
        setTimeout(makeMove, PIECE_ANIMATION_DURATION_MS + TIMEOUT_BUFFER_MS);
    }

    // Start the game loop
    makeMove();
}

